smac/intensification/intensification.py

```diff
@@ -307,7 +163,242 @@ def intensify(self, challengers, incumbent, run_history, aggregate_func,
+                    next_instance = self.rs.choice(list(available_insts))
+                    # Line 7
+                    self.logger.debug("Add run of incumbent")
+                    status, cost, dur, res = self.tae_runner.start(
+                        config=incumbent,
+                        instance=next_instance,
+                        seed=next_seed,
+                        cutoff=self.cutoff,
+                        instance_specific=self.instance_specifics.get(next_instance, "0"))
+
+                    self._num_run += 1
+                else:
+                    self.logger.debug(
+                        "No further instance-seed pairs for incumbent available.")
+                    break
+
+                inc_runs = run_history.get_runs_for_config(incumbent)
+                # Termination condition; after exactly one run, this checks
+                # whether further runs are necessary due to minR
+                if len(inc_runs) >= self.minR or len(inc_runs) >= self.maxR:
+                    break
+
+    def _race_challenger(self, challenger: Configuration, incumbent: Configuration, run_history: RunHistory,
+                         aggregate_func: typing.Callable):
+        '''
+            aggressively race challenger against incumbent
+
+            Parameters
+            ----------
+            challenger : Configuration
+                configuration which challenges incumbent
+            incumbent : Configuration
+                best configuration so far
+            run_history : RunHistory
+                stores all runs we ran so far
+            aggregate_func: typing.Callable
+                aggregate performance across instances
+
+            Returns
+            -------
+            new_incumbent: Configuration
+                either challenger or incumbent
+        '''
+        # at least one run of challenger
+        # to increase chall_indx counter
+        first_run = False
+
+        # Line 8
+        N = max(1, self.minR)
+        
+        inc_inst_seeds = set(run_history.get_runs_for_config(incumbent))
+        # Line 9
+        while True:
+            chall_inst_seeds = set(run_history.get_runs_for_config(challenger))
+            
+            # Line 10
+            missing_runs = list(inc_inst_seeds - chall_inst_seeds)
+
+            # Line 11
+            self.rs.shuffle(missing_runs)
+            to_run = missing_runs[:min(N, len(missing_runs))]
+            # Line 13 (Line 12 comes below...)
+            missing_runs = missing_runs[min(N, len(missing_runs)):]
+
+            inst_seed_pairs = list(inc_inst_seeds - set(missing_runs))
+
+            inc_sum_cost = sum_cost(config=incumbent, instance_seed_pairs=inst_seed_pairs,
+                                    run_history=run_history)
+
+            # Line 12
+            for instance, seed in to_run:
+                # Run challenger on all <config,seed> to run
+                if self.run_obj_time:
```

Comment:
I think this should be an extra function, because it is very specific for runtime optimization and interrupts the reading flow for the racing algorithm.